sdk/batch/batch-rest/src/credentials/batchSharedKeyCredentials.ts

```diff
@@ -0,0 +1,153 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+
+import { type AzureNamedKeyCredential } from "@azure/core-auth";
+import { HttpHeaders, HttpMethods, PipelinePolicy } from "@azure/core-rest-pipeline";
+// import { createHmac } from "crypto";
+
+export function createBatchSharedKeyCredentialsPolicy(
+  // eslint-disable-next-line @azure/azure-sdk/ts-use-interface-parameters
+  credentials: AzureNamedKeyCredential,
+): PipelinePolicy {
+  return {
+    name: "BatchSharedKeyCredentialsPolicy",
+    async sendRequest(request, next) {
+      const accountName = credentials.name;
+      const accountKey = Buffer.from(credentials.key, "base64");
+
+      const ocpDate = request.headers.get("ocp-date");
+      if (!ocpDate) {
+        request.headers.set("ocp-date", new Date().toUTCString());
+      }
+
+      let stringToSign =
+        `${request.method}\n` +
+        getHeaderToAppend(request.headers, "Content-Encoding") +
+        getHeaderToAppend(request.headers, "Content-Language") +
+        getContentLengthToAppend(request.headers, request.method, request.body) +
+        getHeaderToAppend(request.headers, "Content-MD5") +
+        getHeaderToAppend(request.headers, "Content-Type") +
+        getHeaderToAppend(request.headers, "Date") +
+        getHeaderToAppend(request.headers, "If-Modified-Since") +
+        getHeaderToAppend(request.headers, "If-Match") +
+        getHeaderToAppend(request.headers, "If-None-Match") +
+        getHeaderToAppend(request.headers, "If-Unmodified-Since") +
+        getHeaderToAppend(request.headers, "Range");
+
+      // Add canonicalized headers
+      stringToSign += getCanonicalizedHeaders(request.headers);
+
+      // Add canonicalized resource
+      stringToSign += getCanonicalizedResource(request.url, accountName);
+
+      // Signed with sha256
+      try {
+        const crypto = await import("crypto");
```

Comment:
What's the reason of dynamic import here instead of `import { createHmac } from "crypto";` at the top?

Reply:
also, would `computeSha256Hmac()` from @azure/core-util work?

https://github.com/Azure/azure-sdk-for-js/blob/a802c697935f8a22aef904edf18eee0e65f89fcc/sdk/core/core-util/src/sha256.ts#L12

Reply:
The reason for dynamic import is to allow this sdk to work in the browser, as long as the user doesn't use the shared key authentication, which requires the `crypto` module.

Reply:
I tried to use a module replacement like the storage sdk does
https://github.com/Azure/azure-sdk-for-js/blob/main/sdk/storage/storage-blob/package.json#L8

But it did not work for the vitest browser testing.

Reply:
> The reason for dynamic import is to allow this sdk to work in the browser, as long as the user doesn't use the shared key authentication, which requires the `crypto` module.

Since shared key credential is not supported in browsers, you should add a browser version that throw errors if customers attempt to use it.

> But it did not work for the vitest browser testing.

It is only supported in NodeJS so you would only run its test for NodeJS.

Reply:
We still want users to use this package in browser with AAD authentication. The problem is the vitests's browser test won't load the browser version (looks like it doesn't respect the `browser` filed in package.json), it still loads the original version, which will throw an error while importing the `crypto` module.

Reply:
> We still want users to use this package in browser with AAD authentication.

Sure. What I meant is to add a browser version of batchSharedKeyCredentials.ts, so that file gets used in browsers and there would be no importing from "crypto".  For example, storage has this https://github.com/Azure/azure-sdk-for-js/blob/main/sdk/storage/storage-blob/src/credentials/StorageSharedKeyCredential.browser.ts#L4.

> The problem is the vitests's browser test won't load the browser version (looks like it doesn't respect the browser filed in package.json), it still loads the original version, which will throw an error while importing the crypto module.

If the test file that tests shared key credential is nodejs only, you would exclude that test file in the vitest browser config and it should not be executed in browser testing.  Or am I missing anything that you are trying to do?

Reply:
The problem is this import line: 
https://github.com/Azure/azure-sdk-for-js/blob/892fb7c97be57072ad8d689bf6b8d859a046e3f8/sdk/batch/batch-rest/src/batchClient.ts#L8

When running in vitest brower test, it won't import the `batchSharedKeyCredentials.browser.ts` file but import the original `batchSharedKeyCredentials.ts` file, so the test failed, even though I set the `browser` field in package.json. Maybe I was missing some config?

I can give it another try or just mock `batchSharedKeyCredentials.ts` in vitest browser tests.

Reply:
I would expect vitest respect the conditional export and pull in the right files.  https://github.com/Azure/azure-sdk-for-js/blob/892fb7c97be57072ad8d689bf6b8d859a046e3f8/sdk/batch/batch-rest/package.json#L121

We've done similar in a few core packages, for example

https://github.com/Azure/azure-sdk-for-js/blob/main/sdk/core/core-rest-pipeline/src/util/userAgentPlatform-browser.mts#L71

Reply:
Thanks, I added separated files for `batchSharedKeyCredential`. The trick is to name the file ends with `-browser.mts` so that tshy can correctly replace the original file.

Reply:
yeah that's tshy's convention.